src/snowflake/kms.rs

```diff
@@ -193,3 +177,188 @@ impl CryptoMaterialProvider for SnowflakeStageKms {
         Ok(content_material)
     }
 }
+
+#[derive(Clone)]
+pub(crate) struct SnowflakeStageAzureKms {
+    client: Arc<SnowflakeClient>,
+    stage: String,
+    prefix: String,
+    config: SnowflakeStageKmsConfig,
+    keyring: Cache<String, Key>,
+}
+
+impl std::fmt::Debug for SnowflakeStageAzureKms {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SnowflakeStageAzureKms")
+         .field("client", &self.client)
+         .field("stage", &self.stage)
+         .field("config", &self.config)
+         .field("keyring", &"redacted")
+         .finish()
+    }
+}
+
+impl SnowflakeStageAzureKms {
+    pub(crate) fn new(
+        client: Arc<SnowflakeClient>,
+        stage: impl Into<String>,
+        prefix: impl Into<String>,
+        config: SnowflakeStageKmsConfig
+    ) -> SnowflakeStageAzureKms {
+        SnowflakeStageAzureKms {
+            client,
+            stage: stage.into(),
+            prefix: prefix.into(),
+            keyring: Cache::builder()
+                .max_capacity(config.keyring_capacity as u64)
+                .time_to_live(config.keyring_ttl)
+                .build(),
+            config
+        }
+    }
+}
+
+const AZURE_MATDESC_KEY: &str = "matdesc";
+const AZURE_ENCDATA_KEY: &str = "encryptiondata";
+
+#[async_trait::async_trait]
+impl CryptoMaterialProvider for SnowflakeStageAzureKms {
+    async fn material_for_write(&self, _path: &str, data_len: Option<usize>) -> crate::Result<(ContentCryptoMaterial, Attributes)> {
+        let _guard = duration_on_drop!(metrics::material_for_write_duration);
+        let info = self.client.current_upload_info(&self.stage).await?;
+
+        let encryption_material = info.encryption_material.as_ref()
+            .ok_or_else(|| ErrorKind::StorageNotEncrypted(self.stage.clone()))?;
+
+        let description = MaterialDescription {
+            smk_id: encryption_material.smk_id.to_string(),
+            query_id: encryption_material.query_id.clone(),
+            key_size: "128".to_string()
+        };
+        let master_key = Key::from_base64(&encryption_material.query_stage_master_key)
+            .map_err(ErrorKind::MaterialDecode)?;
+
+        let scheme = self.config.crypto_scheme;
+        let material = ContentCryptoMaterial::generate(scheme);
+        let encrypted_cek = material.cek.clone().encrypt_aes_128_ecb(&master_key)
+            .map_err(ErrorKind::MaterialCrypt)?;
```

Comment:
@andrebsguedes I am confused by this. The encryption data says that the key wrapping uses `AES_CBC_256` but the master keys that I am seeing are 128bit. Do you know more about this?